### PR TITLE
Resolves #359: Plan OR predicates with an unordered union

### LINKED ISSUE #359: Plan OR predicates with an unordered union
As part of the new planner work, we want to be able to plan OR queries with union cursors when appropriate. It is always legal to use an `UnorderedUnionPlan`, though it won't remove duplicates even if it can by comparing the children correctly. This is easier to construct the rule for than an ordered `UnionPlan` (which requires that children be compatibly ordered) and is therefore a good first step.

Note that this is something old planner currently does.

### LINKED ISSUE #501: Planner should filter out duplicates with a PrimaryKeyUnorderedDistinctPlan
The new planner currently ignores the `removesDuplicates` parameter included in the `RecordQuery`. It should handle that possibility by adding a `PrimaryKeyUnorderedDistinctPlan` if necessary.

### PR DESCRIPTION
**Pull request by alecgrieser**
This adds a new logical unordered union expression and then rules to create them from a filter and transform the rule into a union.

Because this rule is the first implemented rule (I think) that might create a plan with duplicates and our tests all eschew setting `removes_duplicates` to true, it also made sense to add a thing to remove duplicates with a primary key distinct plan within this work. This included as a separate commit.

This resolves #359, and this resolves #501.

### CODE REVIEW

**Comment by alecgrieser** [comment]
Sorry, I thought I had assigned a reviewer earlier.

I should probably rebase given the changes that have happened since I first submitted (at least to get all the tests to run again).

**Comment by nschiefer** [comment]
Ah, whoops. I had wondered where this change went but was busy enough that I didn't look too hard. Sorry about that. :)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIntersectionPlan.java [hunk 1]
```diff
@@ -210,4 +237,53 @@ public int getComplexity() {
...
+                                                   @Nonnull KeyExpression comparisonKey) {
+        if (left.isReverse() != right.isReverse()) {
+            throw new RecordCoreArgumentException("left plan and right plan for union do not have same value for reverse field");
+        }
+        final List<ExpressionRef<RecordQueryPlan>> childRefs = ImmutableList.of(SingleExpressionRef.of(left), SingleExpressionRef.of(right));
+        return new RecordQueryIntersectionPlan(childRefs, SingleExpressionRef.of(comparisonKey), left.isReverse());
+    }
+
+    /**
+     * Construct a new union of two compatibly-ordered plans. The resulting plan will return all results that are
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIntersectionPlan.java]: "two" -> "two or more"?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryUnionPlan.java [hunk 2]
```diff
@@ -130,4 +163,56 @@ String getDelimiter() {
...
+                                            @Nonnull KeyExpression comparisonKey, boolean showComparisonKey) {
+        if (left.isReverse() != right.isReverse()) {
+            throw new RecordCoreArgumentException("left plan and right plan for union do not have same value for reverse field");
+        }
+        final List<ExpressionRef<RecordQueryPlan>> childRefs = ImmutableList.of(SingleExpressionRef.of(left), SingleExpressionRef.of(right));
+        return new RecordQueryUnionPlan(childRefs, SingleExpressionRef.of(comparisonKey), left.isReverse(), showComparisonKey);
+    }
+
+    /**
+     * Construct a new union of two compatibly-ordered plans. The resulting plan will return all results that are
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryUnionPlan.java]: "two" -> "two or more"?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementUnorderedUnionRule.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * A rule that implements an unordered union of its (already implemented) children. This will combine the
+ * {@link RecordQueryPlan}s corresponding to its children.
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementUnorderedUnionRule.java]: What do you mean "combine the [plans] corresponding to its children"?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Convert a filter on an {@linkplain OrComponent or} expression into a plan on the union. In particular, this will
+ * produce a {@link LogicalUnorderedUnionExpression} with simple filter plans on each child. Those plans can then themselves
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java]: Hmm, the children aren't plans yet. I think that you can just omit the last sentence.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * be planned individually.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class OrToUnorderedUnionRule extends PlannerRule<LogicalFilterExpression> {
+    @Nonnull
+    private static final ReferenceMatcher<QueryComponent> childMatcher = ReferenceMatcher.anyRef();
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java]: Nullability annotations on private static fields like this feel a bit weird (and I didn't use them elsewhere), but I guess I'm not necessarily opposed to them. They do look a bit clunky, though.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * A rule that implements a distinct expression by adding a {@link RecordQueryUnorderedPrimaryKeyDistinctPlan}
+ * if necessary. In particular, it will only add that wrapping expression if the underlying plan itself might
+ * produce duplicate results.
+ *
+ * <p>
+ * This rule is somewhat suspect. In particular, if the inner plan that it matches against does not produce duplicates,
+ * this rule will then return that plan. This is fine unless the plan is later modified in such a way that it then
+ * <em>can</em> produce duplicates. At the moment, none of the rules modify a {@link RecordQueryPlan} once it has
+ * been produced, but a future rule that does so may cause errors where plans erroneously produce duplicate records.
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java]: I think that the standard solution this problem in Cascades is annotating group expressions with required properties, as we discussed offline.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIntersectionPlan.java (hunk 1 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIntersectionPlan.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryUnionPlan.java (hunk 2 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryUnionPlan.java]: Done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java (hunk 6 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java]: I added a comment noting that here as well. I could also add an issue number if there is one.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementUnorderedUnionRule.java (hunk 3 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementUnorderedUnionRule.java]: I tried to reword this to be more exact, though it's still not exactly Edward de Vere.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java (hunk 4 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java]: Okay, just removed that last sentence then.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java (hunk 5 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java]: I could be wrong about this, but I think adding the annotations helps the static analysis tools know whether the expressions can be null or not, which helps it check (for example) if you use one of these as a parameter in a function call, so I try to add them where possible. That may not be exactly correct, though.

**Comment by alecgrieser** [comment]
Rebase is on top of master; comments are responded to in separate commit.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java (hunk 5 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/OrToUnorderedUnionRule.java]: ¯\_(ツ)_/¯ good enough for me.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java (hunk 6 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java]: Created issue #635 for this.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * To address that, the plan is to add a mechanism for enforcing properties (e.g., distinctness or sort order)
+ * on the plans produced by the planner, but that work has not yet been completed.
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java]:
```suggestion
 * on the plans produced by the planner.
 * TODO: Add property tracking/enforcement to group expression (https://github.com/FoundationDB/fdb-record-layer/issues/635)
```

Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/ImplementDistinctRule.java]: I added a "see" link to this. I could also add a TODO, but I think I'd want to add in HTML comment blocks.
